### ISSUE TITLE
0xGoodess - order could still be replayed by multiple honest order sender services in unsync inter-process communication

### ISSUE THREAD

**Issue by sherlock-admin** [sherlock-audit/2023-04-jojo-judging#20]
0xGoodess

high

# order could still be replayed by multiple honest order sender services in unsync inter-process communication

## Summary
order could still be replayed by multiple honest order sender services.

## Vulnerability Detail
While signature replay is specified as not a accepted issue, however the setup of order sender implies there are likely multiple order senders, and multiple such services would easily lead to re-execution due to unsync inter-process communication, if there is no nonce control in the message validation.

`Scenario 1: Dirty usage due to unsync.`

1. Serivce A uses orderSender address (A) to send a taker-order on-chain, on nonce 100. Address A for this taker order would be stored in a db.
2. During the db write this service goes down.
3. Service B now do not know address(A) - nonce100 has been used to send this taker order, so it would pick up this taker order with another order sender address (B) and send it on-chain, using B's nonce 10.
4. the order is executed twice under nonce 100 from (A) and nonce 10 from (B). 

`Scenario 2 - Operation Risk`

1. Serivce A uses orderSender address (A) to send a taker-order on-chain, on nonce 100. 
2. Gas spikes and the address (A) does not have enough fee to execute, the tx stuck on the mempool. (or the Arbitrium sequencer cache)
3. Service B with Address (B) takes the tx and send it on-chain
4. address A gets replenished with gas now and the original tx gets executed immediately.

`Scenario 3 - Arbitrium sequencer becomes decentralized.`

1. In a decentralized environment, there would be a chance of re-org.
2. Let's say orderSender A sends a taker order on-chain (under nonce 100), the tx is confirmed
3. Arbitrium reorg and the backend realizes the tx needs to be resent
4. orderSender B (due to roundRobin or any other loadbalancing algo) sends the same taker order (under its nonce)
5. The tx gets executed twice.

In short, having no nonce validation puts much constraint on the backend as well as limiting the thorough-put of orderSender. On the contrary the cost of implementing it is just an additional check.

## Impact
Re-execution the same order on-chain even though backend has a nonce control.

## Code Snippet
https://github.com/sherlock-audit/2023-04-jojo/blob/main/smart-contract-EVM/contracts/impl/JOJOExternal.sol#L103-L163

## Tool used

Manual Review

## Recommendation

implement a nonce check on the `approveTrade` function.

```solidity
require(state.nonce[trader] == order.nonce);
state.nonce[trader]++;
```

Duplicate of #414
